### PR TITLE
Make `randomTimeValue()` return a `TimeValue`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today various test utility methods construct a random string that can be
parsed as a `TimeValue`, but in practice almost everywhere we just parse
it straight away. With this change we have the utility methods return a
`TimeValue` directly.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Sorry for the extra iterations here, this is good to review now.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/DownsampleStepTests.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
 
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.TimeUnit;
```
Comment (pgomulka) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/DownsampleStepTests.java:29]: perhaps java.time.temporal.ChronoUnit would be better ?

Comment (pgomulka) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/DownsampleStepTests.java:29]: actually it is already being used in ESTestCase so would make the PR far too big.

**Review by pgomulka — APPROVED** [review]
LGTM,
it was slightly more complex than I expected. I suspected it wasn't a 'mechanical refactoring' ?
I glimpsed at all files and they look good to me

**Comment by DaveCTurner** [comment]
Somewhat mechanical but then a bunch of manual cleanup needed too.

**Review by DiannaHohensee — APPROVED** [review]
Noice 😎 Left a few suggestions. Great to remove all that parsing.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/DownsampleStepTests.java (hunk 1 above)
Comment (DaveCTurner) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/DownsampleStepTests.java:29]: Yeah that'd be quite a big change, `ChronoUnit` is a lot more general than `TimeUnit` but we wouldn't want to add support for all those other units in our APIs today.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/ActionConfigStatsTests.java [hunk 2]
```diff
@@ -45,7 +45,7 @@ public static ActionConfigStats createRandomInstance() {
             builder.setRolloverMaxSize(randomByteSize());
         }
         if (randomBoolean()) {
-            TimeValue randomAge = TimeValue.parseTimeValue(randomTimeValue(), "action_config_stats_tests");
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/ActionConfigStatsTests.java:48]: ++ 2373d2f3f6e81c3c07b983afd09461d76301bfb1

File: server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java [hunk 3]
```diff
@@ -637,11 +639,7 @@ public void testRemainingDelayCalculationsWithUnrelatedShutdowns() {
                     .setStartedAtMillis(randomNonNegativeLong())
                     .setType(type)
                     .setTargetNodeName(targetNodeName)
-                    .setGracePeriod(
-                        type == SingleNodeShutdownMetadata.Type.SIGTERM
-                            ? TimeValue.parseTimeValue(randomTimeValue(), this.getTestName())
-                            : null
-                    )
+                    .setGracePeriod(type == SingleNodeShutdownMetadata.Type.SIGTERM ? randomTimeValue() : null)
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/UnassignedInfoTests.java]: I didn't really mean to do anything with it, normally I just let IntelliJ take the lead. But you're right it didn't make the most intellijent decisions here. 4535cf5efe66c672b14c1e324a5a22521e4b1b54

File: server/src/test/java/org/elasticsearch/script/ScriptServiceTests.java [hunk 4]
```diff
@@ -554,18 +555,16 @@ public void testFallbackContextSettings() {
         int cacheSizeBackup = randomIntBetween(0, 1024);
         int cacheSizeFoo = randomValueOtherThan(cacheSizeBackup, () -> randomIntBetween(0, 1024));
 
-        String cacheExpireBackup = randomTimeValue(1, 1000, "h");
-        TimeValue cacheExpireBackupParsed = TimeValue.parseTimeValue(cacheExpireBackup, "");
-        String cacheExpireFoo = randomValueOtherThan(cacheExpireBackup, () -> randomTimeValue(1, 1000, "h"));
-        TimeValue cacheExpireFooParsed = TimeValue.parseTimeValue(cacheExpireFoo, "");
+        TimeValue cacheExpireBackupParsed = randomTimeValue(1, 1000, TimeUnit.HOURS);
+        TimeValue cacheExpireFooParsed = randomValueOtherThan(cacheExpireBackupParsed, () -> randomTimeValue(1, 1000, TimeUnit.HOURS));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/script/ScriptServiceTests.java]: ++ e323a19fea13036a97a805c2f7d59cf995fc734f

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/transform/transforms/pivot/DateHistogramGroupSourceTests.java [hunk 5]
```diff
@@ -84,7 +86,7 @@ public static DateHistogramGroupSource randomDateHistogramGroupSource(
                 scriptConfig,
                 missingBucket,
                 new DateHistogramGroupSource.CalendarInterval(
-                    new DateHistogramInterval(randomTimeValue(1, 1, "m", "h", "d", "w", "M", "q", "y"))
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/transform/transforms/pivot/DateHistogramGroupSourceTests.java:87]: Yeah this was kind of abuse of the `TimeValue` name, we support much more general intervals in aggs than we do in most APIs.

File: x-pack/plugin/shutdown/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/shutdown/NodeShutdownIT.java [hunk 6]
```diff
@@ -96,6 +98,10 @@ public void checkCRUD(
         }
     }
 
+    private static Matcher<Object> equalsOptionalTimeValue(TimeValue timeValue) {
+        return timeValue == null ? nullValue() : equalTo(timeValue.getStringRep());
+    }
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/shutdown/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/shutdown/NodeShutdownIT.java:103]: In this suite we already start with test methods, followed by the utilities on which they depend. My hunch is that this order is more common, but it's certainly not very consistent. Again I tend to just follow whatever IntelliJ suggests...

File: x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlQueryRequestTests.java [hunk 7]
```diff
@@ -172,7 +172,7 @@ private RequestInfo randomRequestInfo() {
     }
 
     private TimeValue randomTV() {
-        return TimeValue.parseTimeValue(randomTimeValue(), null, "test");
+        return randomTimeValue();
```
Comment (ywangd) [inline x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlQueryRequestTests.java:175]: This method does not seem to be useful after this refactor

File: x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlTranslateRequestTests.java [hunk 8]
```diff
@@ -63,7 +63,7 @@ protected Writeable.Reader<TestSqlTranslateRequest> instanceReader() {
     }
 
     private TimeValue randomTV() {
-        return TimeValue.parseTimeValue(randomTimeValue(), null, "test");
+        return randomTimeValue();
```
Comment (ywangd) [inline x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlTranslateRequestTests.java:66]: Same here

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/common/http/HttpRequestTests.java [hunk 9]
```diff
@@ -105,16 +106,11 @@ public void testXContentSerialization() throws Exception {
         }
         if (randomBoolean()) {
             // micros and nanos don't round trip will full precision so exclude them from the test
-            String safeConnectionTimeout = randomValueOtherThanMany(
-                s -> (s.endsWith("micros") || s.endsWith("nanos")),
-                () -> randomTimeValue()
-            );
-            builder.connectionTimeout(TimeValue.parseTimeValue(safeConnectionTimeout, "my.setting"));
+            builder.connectionTimeout(randomTimeValue(0, 1000, TimeUnit.DAYS, TimeUnit.HOURS, TimeUnit.MINUTES, TimeUnit.SECONDS));
```
Comment (ywangd) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/common/http/HttpRequestTests.java:109]: Based on the existing code and comment, it seems we should also include `TimeUnit.MILLISECONDS`

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/common/http/HttpRequestTests.java [hunk 10]
```diff
@@ -105,16 +106,11 @@ public void testXContentSerialization() throws Exception {
...
         }
         if (randomBoolean()) {
             // micros and nanos don't round trip will full precision so exclude them from the test
-            String safeReadTimeout = randomValueOtherThanMany(s -> (s.endsWith("micros") || s.endsWith("nanos")), () -> randomTimeValue());
-            builder.readTimeout(TimeValue.parseTimeValue(safeReadTimeout, "my.setting"));
+            builder.readTimeout(randomTimeValue(0, 1000, TimeUnit.DAYS, TimeUnit.HOURS, TimeUnit.MINUTES, TimeUnit.SECONDS));
```
Comment (ywangd) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/common/http/HttpRequestTests.java:113]: Same here

**Review by ywangd — APPROVED** [review]
LGTM

**Comment by ywangd** [comment]
Will this change break stateless build?

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/common/http/HttpRequestTests.java (hunk 9 above)
Comment (DaveCTurner) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/common/http/HttpRequestTests.java]: Ah thanks, missed that.

File: x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlTranslateRequestTests.java (hunk 8 above)
Comment (DaveCTurner) [reply x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlTranslateRequestTests.java]: Good point, see 45333d2e6f2528e44238f61d13aaf1da483b1de8.

File: x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlQueryRequestTests.java (hunk 7 above)
Comment (DaveCTurner) [reply x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlQueryRequestTests.java]: Good point, see 45333d2e6f2528e44238f61d13aaf1da483b1de8.

**Comment by DaveCTurner** [comment]
> Will this change break stateless build?

Hardly at all 🙂  I opened a PR to handle it.

**Pull request merged**